runtime/autoload/man.vim

```diff
@@ -2,14 +2,84 @@
 
 let s:man_find_arg = "-w"
 
+" Handler for s:system() function.
+function! s:system_handler(jobid, data, event) dict abort
+  if a:event is# 'stdout' || a:event is# 'stderr'
+    let self[a:event] .= join(a:data, "\n")
+  else
+    let self.exit_code = a:data
+  endif
+endfunction
+
+" Run a system command and timeout after 30 seconds.
+function! s:system(cmd, ...) abort
+  let opts = {
+        \ 'stdout': '',
+        \ 'stderr': '',
+        \ 'exit_code': 0,
+        \ 'on_stdout': function('s:system_handler'),
+        \ 'on_stderr': function('s:system_handler'),
+        \ 'on_exit': function('s:system_handler'),
+        \ }
+  let jobid = jobstart(a:cmd, opts)
+
+  if jobid < 1
+    throw printf('command error %d: %s', jobid, join(a:cmd))
+  endif
+
+  let res = jobwait([jobid], 30000)
+  if res[0] == -1
+    try
+      call jobstop(jobid)
+      throw printf('command timed out: %s', join(a:cmd))
+    catch /^Vim(call):E900:/
+    endtry
+  elseif res[0] == -2
+    throw printf('command interrupted: %s', join(a:cmd))
+  endif
+  if opts.exit_code != 0
+    throw printf("command error (%d) %s: %s", jobid, join(a:cmd), substitute(opts.stderr, '\_s\+$', '', &gdefault ? '' : 'g'))
+  endif
+
+  return opts.stdout
+endfunction
+
+function! s:get_path(sect, name) abort
+  if empty(a:sect)
+    return s:system(['man', s:man_find_arg, a:name])
+  endif
+  " '-s' flag handles:
+  "   - tokens like 'printf(echo)'
+  "   - sections starting with '-'
+  "   - 3pcap section (found on macOS)
+  "   - commas between sections (for section priority)
+  return s:system(['man', s:man_find_arg, '-s', a:sect, a:name])
+endfunction
+
+function! s:get_page(path) abort
+  " Respect $MANWIDTH or default to window width.
+  let manwidth = empty($MANWIDTH) ? winwidth(0) : $MANWIDTH
+  " Force MANPAGER=cat to ensure Vim is not recursively invoked (by man-db).
+  " http://comments.gmane.org/gmane.editors.vim.devel/29085
+  let cmd = ['env', 'MANPAGER=cat', 'MANWIDTH='.manwidth, 'man']
+  return s:system(cmd + (s:man_has_l_flag ? ['-l', a:path] : [a:path]))
+endfunction
+
 " TODO(nhooyr) Completion may work on SunOS; I'm not sure if `man -l` displays
 " the list of searched directories.
 try
   if !has('win32') && $OSTYPE !~? 'cygwin\|linux' && system('uname -s') =~? 'SunOS' && system('uname -r') =~# '^5'
     let s:man_find_arg = '-l'
   endif
+  let s:man_has_l_flag = 1
+  call s:get_page(s:get_path('', 'man'))
 catch /E145:/
   " Ignore the error in restricted mode
+catch /command error .*/
+  let s:man_has_l_flag = 0
+catch
+  let s:man_has_l_flag = 0
+  echomsg 'man.vim: ' . v:exception
 endtry
```

Comment:
Moved the initialization code into `s:init` to avoid future code churn. I hope this is enough.